GitUI/CommitInfo/CommitInfo.cs

```diff
@@ -801,20 +801,99 @@ protected override void DisposeCustomResources()
-                RegexOptions.Compiled | RegexOptions.CultureInvariant);
-
+            private const string _remoteBranchPrefix = "remotes/";
             private readonly string _currentBranch;
             private readonly bool _isDetachedHead;
+            private Dictionary<string, int> _orderByBranch = new();
 
-            public BranchComparer(string currentBranch)
+            public BranchComparer(List<string> branches, string currentBranch)
             {
                 _currentBranch = currentBranch;
                 _isDetachedHead = DetachedHeadParser.IsDetachedHead(currentBranch);
+                string[] branchRegexes = AppSettings.PrioritizedBranchNames.Split(";", StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
+                string[] localBranchRegexes = branchRegexes.Select(regex => $"^({regex})$").ToArray();
+                string[] remoteBranchRegexes = branchRegexes.Select(regex => $"^{_remoteBranchPrefix}[^/]+/({regex})$").ToArray();
+                string[] remoteRegexes = AppSettings.PrioritizedRemoteNames.Split(";", StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
+                    .Select(regex => $"^{_remoteBranchPrefix}({regex})/").ToArray();
+
+                foreach (string branch in branches)
+                {
+                    _orderByBranch[branch] = GetBranchOrder(branch);
+                }
+
+                return;
+
+                // Get the order for each branch.
+                // Add max possible order value to next "level" to sort properly with the order for each regex.
+                int GetBranchOrder(string branch)
+                {
+                    int order = 0;
+                    if (_isDetachedHead ? DetachedHeadParser.IsDetachedHead(branch) : branch == _currentBranch)
+                    {
+                        return order;
+                    }
+
+                    order++;
+                    if (IsLocalBranch())
+                    {
+                        if (!TryGetOrder(branch, localBranchRegexes, out int localBranchOrder))
+                        {
+                            // Non prioritized branches added after prioritized remote branches
+                            order += prioritizedRemoteBranchesLength();
+                        }
+
+                        // Order by branch priority
+                        // Non priority has localBranchOrder==localBranchRegexes.Length, an extra priority level
+                        order += localBranchOrder;
+
+                        return order;
+                    }
+
+                    // Remote branches
+                    order += localBranchRegexes.Length;
+                    if (!TryGetOrder(branch, remoteBranchRegexes, out int remoteBranchOrder))
+                    {
+                        // after non priority local branches
+                        order += 1;
```

Comment:
nit
```suggestion
                        order++;
```
although I prefer the prefix increment operator `++order` because of runtime costs with e.g. iterators (in C++, which should rather have been named ++C), and it reads "Increment variable" like functions.

Reply:
This is intentional, to have the pattern "order +=", it just happen to be one here

...and C++ is a better name than ++C, it is next language that will be better. That is the price for backwards compatibility.

Reply:
Can add a constant for the 1 and do the same after 'current branch' check